### ISSUE TITLE
[1.12] crash after OOS

### ISSUE THREAD

**Issue by sevu** [wesnoth/wesnoth#2593]
I had a crash in 1.12, and the coredump gave me the following information:

```
Using host libthread_db library "/usr/lib/libthread_db.so.1".
Core was generated by `wesnoth'.
Program terminated with signal SIGABRT, Aborted.
#0  0x00007f0fadf2b860 in raise () from /usr/lib/libc.so.6
[Current thread is 1 (Thread 0x7f0fb1f229c0 (LWP 6666))]
(gdb) bt
#0  0x00007f0fadf2b860 in raise () at /usr/lib/libc.so.6
#1  0x00007f0fadf2cec9 in abort () at /usr/lib/libc.so.6
#2  0x00007f0fadf240bc in __assert_fail_base () at /usr/lib/libc.so.6
#3  0x00007f0fadf24133 in  () at /usr/lib/libc.so.6
#4  0x000055ddcbc4d122 in playmp_controller::handle_generic_event(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) ()
#5  0x000055ddcba517ab in events::generic_event::notify_observers() ()
#6  0x000055ddcbcbbcc4 in synced_context::pull_remote_user_input(bool) ()
#7  0x000055ddcbc64a76 in  ()
#8  0x000055ddcbc66570 in mp_sync::get_user_choice(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, mp_sync::user_choice const&, int) ()
#9  0x000055ddcb9fd3d4 in  ()
#10 0x000055ddcbc7d40a in  ()
#11 0x000055ddcbe1dea8 in  ()
#12 0x000055ddcbe2d6b2 in  ()
#13 0x000055ddcbe1e60f in  ()
#14 0x000055ddcbe1f128 in  ()
#15 0x000055ddcbe1b3ea in lua_pcallk(lua_State*, int, int, int, int, int (*)(lua_State*)) ()
#16 0x000055ddcbc94cb7 in luaW_pcall(lua_State*, int, int, bool) ()
#17 0x000055ddcbc86957 in LuaKernel::run_wml_action(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, vconfig const&, game_events::queued_event const&) ()
#18 0x000055ddcba1a34a in game_events::handle_event_commands(game_events::queued_event const&, vconfig const&) ()
#19 0x000055ddcba0fcde in game_events::event_handler::handle_event(game_events::queued_event const&, boost::shared_ptr<game_events::event_handler>&) ()
#20 0x000055ddcba1ea14 in game_events::pump() ()
#21 0x000055ddcba1f81a in game_events::fire(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, game_events::entity_location const&, game_events::entity_location const&, config const&) ()
#22 0x000055ddcb7849af in  ()
#23 0x000055ddcb786acc in actions::move_unit_from_replay(std::vector<map_location, std::allocator<map_location> > const&, actions::undo_list*, bool, bool, bool) ()
#24 0x000055ddcbcc08a1 in  ()
#25 0x000055ddcbcbe347 in synced_context::run_in_synced_context(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, config const&, bool, bool, bool, boost::function2<void, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool>) ()
#26 0x000055ddcbc633ec in do_replay_handle(int) ()
#27 0x000055ddcbc639a4 in do_replay(int) ()
#28 0x000055ddcbc575c5 in turn_info::process_network_data(config const&, bool) ()
#29 0x000055ddcbc4bbfb in playmp_controller::play_human_turn() ()
#30 0x000055ddcbc4d8d1 in playmp_controller::linger() ()
#31 0x000055ddcbc47bec in play_game(game_display&, game_state&, config const&, io_type_t, bool, bool, bool) ()
#32 0x000055ddcbbab683 in  ()
#33 0x000055ddcbbabc6b in  ()
#34 0x000055ddcbbac28e in mp::start_client(game_display&, config const&, game_state&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) ()
#35 0x000055ddcb9db994 in game_controller::play_multiplayer() ()
#36 0x000055ddcb6150d2 in  ()
#37 0x000055ddcb5f41c4 in main ()
```
As for the reason of the crash - I played the planewalker MP campaign, and there seems to be a bug in the campaign during switching to the next level when not triggering the last event as side 1/host.
The first time I ran in this UMC bug the game just didn't end.
The 2nd time I got an OOS followed by an crash, before pressing a button. I think the other player left or changed the controller of my side while this happened.

Long story short - Is the BT useful for anything?

**Comment by gfgtdf** [comment]
well the stacktrace gives an assertino error. the mesage of it would be helpful.


Other than that, the code (https://github.com/wesnoth/wesnoth/blob/1.13.11/src/playmp_controller.cpp#L440 for the 1.13 equivalent) looks a bit like it'd assert on network input which is clearly a bad thing to do

**Comment by gfgtdf** [comment]
ok after looking at it a second time sync_network can really not return PROCESS_END_TURN in that situation it could return PROCESS_END_LINGER if somethign really went wrong though.

**Comment by sevu** [comment]
Maybe it did. I was in linger_mode, although the scenario should have ended / the selected one loaded.
For the other player it looked like I didn't end the turn, and when changing the controller there was a message that some actions which haven't been sent to the other player have been removed.

The stack trace looked like that:
```
                Stack trace of thread 6666:
                #0  0x00007f0fadf2b860 raise (libc.so.6)
                #1  0x00007f0fadf2cec9 abort (libc.so.6)
                #2  0x00007f0fadf240bc __assert_fail_base (libc.so.6)
                #3  0x00007f0fadf24133 __assert_fail (libc.so.6)
                #4  0x000055ddcbc4d122 _ZN17playmp_controller20handle_generic_eventERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE (wesnoth)
                #5  0x000055ddcba517ab _ZN6events13generic_event16notify_observersEv (wesnoth)
                #6  0x000055ddcbcbbcc4 _ZN14synced_context22pull_remote_user_inputEb (wesnoth)
                #7  0x000055ddcbc64a76 n/a (wesnoth)
                #8  0x000055ddcbc66570 _ZN7mp_sync15get_user_choiceERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERKNS_11user_choiceEi (wesnoth)
                #9  0x000055ddcb9fd3d4 n/a (wesnoth)
                #10 0x000055ddcbc7d40a n/a (wesnoth)
                #11 0x000055ddcbe1dea8 n/a (wesnoth)
                #12 0x000055ddcbe2d6b2 n/a (wesnoth)
                #13 0x000055ddcbe1e60f n/a (wesnoth)
                #14 0x000055ddcbe1f128 n/a (wesnoth)
                #15 0x000055ddcbe1b3ea _Z10lua_pcallkP9lua_StateiiiiPFiS0_E (wesnoth)
                #16 0x000055ddcbc94cb7 _Z10luaW_pcallP9lua_Stateiib (wesnoth)
                #17 0x000055ddcbc86957 _ZN9LuaKernel14run_wml_actionERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERK7vconfigRKN11game_events12queued_eventE (wesnoth)
                #18 0x000055ddcba1a34a _ZN11game_events21handle_event_commandsERKNS_12queued_eventERK7vconfig (wesnoth)
                #19 0x000055ddcba0fcde _ZN11game_events13event_handler12handle_eventERKNS_12queued_eventERN5boost10shared_ptrIS0_EE (wesnoth)
                #20 0x000055ddcba1ea14 _ZN11game_events4pumpEv (wesnoth)
                #21 0x000055ddcba1f81a _ZN11game_events4fireERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERKNS_15entity_locationESA_RK6config (wesnoth)
                #22 0x000055ddcb7849af n/a (wesnoth)
                #23 0x000055ddcb786acc _ZN7actions21move_unit_from_replayERKSt6vectorI12map_locationSaIS1_EEPNS_9undo_listEbbb (wesnoth)
                #24 0x000055ddcbcc08a1 n/a (wesnoth)
                #25 0x000055ddcbcbe347 _ZN14synced_context21run_in_synced_contextERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERK6configbbbN5boost9function2IvS7_bEE (wesnoth)
                #26 0x000055ddcbc633ec _Z16do_replay_handlei (wesnoth)
                #27 0x000055ddcbc639a4 _Z9do_replayi (wesnoth)
                #28 0x000055ddcbc575c5 _ZN9turn_info20process_network_dataERK6configb (wesnoth)
                #29 0x000055ddcbc4bbfb _ZN17playmp_controller15play_human_turnEv (wesnoth)
                #30 0x000055ddcbc4d8d1 _ZN17playmp_controller6lingerEv (wesnoth)
                #31 0x000055ddcbc47bec _Z9play_gameR12game_displayR10game_stateRK6config9io_type_tbbb (wesnoth)
                #32 0x000055ddcbbab683 n/a (wesnoth)
                #33 0x000055ddcbbabc6b n/a (wesnoth)
                #34 0x000055ddcbbac28e _ZN2mp12start_clientER12game_displayRK6configR10game_stateRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE (wesnoth)
                #35 0x000055ddcb9db994 _ZN15game_controller16play_multiplayerEv (wesnoth)
                #36 0x000055ddcb6150d2 n/a (wesnoth)
                #37 0x000055ddcb5f41c4 main (wesnoth)
                #38 0x00007f0fadf17f4a __libc_start_main (libc.so.6)
                #39 0x000055ddcb61142a _start (wesnoth)
                
                Stack trace of thread 6676:
                #0  0x00007f0fb18bf260 __nanosleep (libpthread.so.0)
                #1  0x00007f0fb1b1fa46 SDL_Delay (libSDL-1.2.so.0)
                #2  0x00007f0fb1b1fa93 n/a (libSDL-1.2.so.0)
                #3  0x00007f0fb1adc00b n/a (libSDL-1.2.so.0)
                #4  0x00007f0fb1b1d30a n/a (libSDL-1.2.so.0)
                #5  0x00007f0fb18b508c start_thread (libpthread.so.0)
                #6  0x00007f0fadfece7f __clone (libc.so.6)
                
                Stack trace of thread 6686:
                #0  0x00007f0fadfe2a76 ppoll (libc.so.6)
                #1  0x00007f0fa5d50c53 pa_mainloop_poll (libpulse.so.0)
                #2  0x00007f0fa5d51271 pa_mainloop_iterate (libpulse.so.0)
                #3  0x00007f0fb1aff20d n/a (libSDL-1.2.so.0)
                #4  0x00007f0fb1ad3994 n/a (libSDL-1.2.so.0)
                #5  0x00007f0fb1adc00b n/a (libSDL-1.2.so.0)
                #6  0x00007f0fb1b1d30a n/a (libSDL-1.2.so.0)
                #7  0x00007f0fb18b508c start_thread (libpthread.so.0)
                #8  0x00007f0fadfece7f __clone (libc.so.6)
                
                Stack trace of thread 6691:
                #0  0x00007f0fb18bb3bd pthread_cond_wait@@GLIBC_2.3.2 (libpthread.so.0)
                #1  0x00007f0fb1b1db9f SDL_CondWait (libSDL-1.2.so.0)
                #2  0x000055ddcbdd3100 _ZN9threading9condition4waitERKNS_5mutexE (wesnoth)
                #3  0x000055ddcbda782b n/a (wesnoth)
                #4  0x00007f0fb1adc00b n/a (libSDL-1.2.so.0)
                #5  0x00007f0fb1b1d30a n/a (libSDL-1.2.so.0)
                #6  0x00007f0fb18b508c start_thread (libpthread.so.0)
                #7  0x00007f0fadfece7f __clone (libc.so.6)
                
                Stack trace of thread 7382:
                #0  0x00007f0fae4ddf0e futex_wait (libgomp.so.1)
                #1  0x00007f0fae4db8c8 gomp_simple_barrier_wait (libgomp.so.1)
                #2  0x00007f0fb18b508c start_thread (libpthread.so.0)
                #3  0x00007f0fadfece7f __clone (libc.so.6)
                
                Stack trace of thread 7383:
                #0  0x00007f0fae4ddf0e futex_wait (libgomp.so.1)
                #1  0x00007f0fae4db8c8 gomp_simple_barrier_wait (libgomp.so.1)
                #2  0x00007f0fb18b508c start_thread (libpthread.so.0)
                #3  0x00007f0fadfece7f __clone (libc.so.6)
                
                Stack trace of thread 7384:
                #0  0x00007f0fae4ddf0e futex_wait (libgomp.so.1)
                #1  0x00007f0fae4db8c8 gomp_simple_barrier_wait (libgomp.so.1)
                #2  0x00007f0fb18b508c start_thread (libpthread.so.0)
                #3  0x00007f0fadfece7f __clone (libc.so.6)
```
I looked how this event works in the add-on, but couldn't see anything which would work different if player 2 / not the host triggers the event to choose the next level.

**Comment by gfgtdf** [comment]
hmm ye it somehow still doesn't tell the assertion erro message, usualyl assertion failures come with a message thal tell you which assertion exactly failed.


anyways, i changed the code so that it now no longer asserts on netowkr input https://github.com/wesnoth/wesnoth/commit/18ee8b767c155999e0160f3d4bfd88f603796d32, it will now throw you to the lobby with a network error message instead in that (rare) care.

**Comment by gfgtdf** [comment]
you still didn't give the asserts message so i cannot be 100% sure, but i'd say ma commit above fixed it
